In [1]:
#importing libraries
import pandas as pd 
import numpy as np
import os

In [2]:
#creating path
path=r'C:\Users\akul_\OneDrive\Documents\Tarishi\CF Data analysis bootcamp\Achievement 4-Python\Instacart Basket Analysis'

In [3]:
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'))

In [4]:
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_wrangled.csv'))

In [5]:
df_ords.describe()

Unnamed: 0      order_id       user_id  no_of_orders  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [6]:
#create a new dataframe for testing
df_test=pd.DataFrame()

In [7]:
#create a mixed type column
df_test['mix']=['a','b',1,True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
#check for mixed type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\akul_\AppData\Local\Temp\ipykernel_15656\1479560198.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [10]:
# to check for missing values: if any null value, it would add True/1 else false/0, in this case 16 null values in prodect_name
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
# create a new clean datafram with only non missing values 
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
# check for duplicate data 
df_prods_dups=df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_prods_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
# create a new dataframe without duplicates
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
#exporting data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared data', 'products_checked.csv'), index=False)

# Task 4.5

# df_ords consistnecy check

In [21]:
#01 df_ords descriptives
df_ords.describe()

Unnamed: 0      order_id       user_id  no_of_orders  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# there is an unnamed column in the dataframe, which I am not sure of. 
# the count of all the columns should be same but it seems like there are some missing values in the days_since_prior_order column as the count is not same. 

In [22]:
#02 to check for mixed data type
for col in df_ords.columns:
    col_types = df_ords[col].map(type)
    if col_types.nunique() > 1:
        print(f"Column '{col}' has mixed data types")
    else:
        print(f"no mixed data in '{col}'")

no mixed data in 'Unnamed: 0'
no mixed data in 'order_id'
no mixed data in 'user_id'
no mixed data in 'no_of_orders'
no mixed data in 'orders_day_of_week'
no mixed data in 'order_hour_of_day'
no mixed data in 'days_since_prior_order'


In [23]:
print(df_ords.dtypes)

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
no_of_orders                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object


In [24]:
#03 check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
no_of_orders                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# There are 206209 missing values for days_since_prior_order. A subset is created to view these missing values and incestigate further.

In [25]:
#creating a new subset to view these missing values
df_ords_miss=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [26]:
df_ords_miss

Unnamed: 0  order_id  user_id  no_of_orders  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

# The missing values in the days_since_prior_order column are left as it is. The missing value in the column means that there have been 0 days before the last order was placed. This is crucial to understanding the purchasing frequency of the customer. 

In [27]:
#04 check for duplicates
df_ords_dups=df_ords[df_ords.duplicated()]


In [28]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, no_of_orders, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [29]:
df_ords.duplicated().sum()

0

# there are no duplicate records found


In [30]:
df_ords.shape

(3421083, 7)

In [31]:
# exporting the files
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared data', 'orders_checked.csv'))